cpp/src/arrow/table.h

```diff
@@ -23,6 +23,7 @@
 #include <vector>
 
 #include "arrow/chunked_array.h"  // IWYU pragma: keep
+#include "arrow/compute/cast.h"
```

Comment:
It does not seem to find the class when building the tests:
```
/usr/bin/ld: debug/libarrow.so.1300.0.0: undefined reference to `arrow::PromoteTableToSchema(std::shared_ptr<arrow::Table> const&, std::shared_ptr<arrow::Schema> const&, arrow::compute::CastOptions const&, arrow::MemoryPool*)'
collect2: error: ld returned 1 exit status
[319/835] Building CXX object src/arrow/compute/kernels/CMakeFiles/arrow-compute-scalar-if-else-test.dir/scalar_if_else_test.cc.o
ninja: build stopped: subcommand failed.
```
Any pointers?

Also, I think it would be nice to set the `CastOptions` to `::Safe()` by default, but I'm not able to get this working with the forward reference.

Reply:
It's the _function_ `PromoteTableToSchema` that fails linking in your error message above. Meaning you probably didn't implement that particular signature?

> Also, I think it would be nice to set the `CastOptions` to `::Safe()` by default, but I'm not able to get this working with the forward reference.

Yes, this is why I suggested two signatures, one with `CastOptions` and one without.

Reply:
Got it, thanks. I changed the `CastOptions` to a pointer because I can't forward declare the `CastOptions::Safe()` afaik. If there is a nullpointer, it will be set to `Safe()` within the `#ifdef ARROW_COMPUTE` block. Let me know what you think

Reply:
About including the compute headers here, @pitrou are you fine with the current state, or do you have another suggestion how to avoid it?

Reply:
Still no fine, and the suggestion I made in https://github.com/apache/arrow/pull/36846/files/1c5643c4a855ca7a29fdbc4e3b83acd4e3d2af96#r1290117358 still holds.

Reply:
Removed the import 👍  Digging into the failing CI